File: backend/Ember.Domain/Data/Invitation.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace Ember.Domain.Data;

/// <summary>
/// Represents an invitation sent by an existing Ember user to allow a new person to register.
/// The invited person must accept the invitation to create a new EmberUser account.
/// </summary>
public sealed class Invitation
{
    [Key]
    public Guid Id { get; set; }

    [Required]
    public required Guid InvitedByUserId { get; set; }

    [ForeignKey(nameof(InvitedByUserId))]
    public EmberUser? InvitedByUser { get; set; }

    // The real name provided when creating the invitation
    [Required]
    public required string RealName { get; set; }

    // True when the invited person is confirmed to be of legal age
    public bool IsInLegalAge { get; set; }

    // Jurisdiction provided for the invited person
    [Required]
    public required string Jurisdiction { get; set; }

    // Contact information for the invited person (email or phone or both)
    public string? Email { get; set; }
    public string? Phone { get; set; }

    // A unique code/token used to accept the invitation
    [Required]
    public required string InviteCode { get; set; }

    public required DateTimeOffset CreatedAt { get; set; }
    public DateTimeOffset? ExpiresAt { get; set; }

    // When set, indicates the invitation was accepted and a user was created
    public DateTimeOffset? AcceptedAt { get; set; }

    // If accepted, link to the created user
    public Guid? AcceptedByUserId { get; set; }

    [ForeignKey(nameof(AcceptedByUserId))]
    public EmberUser? AcceptedByUser { get; set; }
}

File: flutter_app/lib/models/invitation.dart
class Invitation {
  final String id;
  final String inviteCode;
  final String realName;
  final bool isInLegalAge;
  final String jurisdiction;
  final String? email;
  final String? phone;
  final DateTime createdAt;
  final DateTime? expiresAt;
  final DateTime? acceptedAt;
  final String? acceptedByUserId;

  const Invitation({
    required this.id,
    required this.inviteCode,
    required this.realName,
    required this.isInLegalAge,
    required this.jurisdiction,
    this.email,
    this.phone,
    required this.createdAt,
    this.expiresAt,
    this.acceptedAt,
    this.acceptedByUserId,
  });

  bool get isAccepted => acceptedAt != null;
  bool get isExpired =>
      expiresAt != null && expiresAt!.isBefore(DateTime.now().toUtc());
  bool get isPending => !isAccepted && !isExpired;

  factory Invitation.fromJson(Map<String, dynamic> json) {
    return Invitation(
      id: json['id'] as String,
      inviteCode: json['inviteCode'] as String,
      realName: json['realName'] as String,
      isInLegalAge: json['isInLegalAge'] as bool,
      jurisdiction: json['jurisdiction'] as String,
      email: json['email'] as String?,
      phone: json['phone'] as String?,
      createdAt: DateTime.parse(json['createdAt'] as String),
      expiresAt: json['expiresAt'] != null
          ? DateTime.parse(json['expiresAt'] as String)
          : null,
      acceptedAt: json['acceptedAt'] != null
          ? DateTime.parse(json['acceptedAt'] as String)
          : null,
      acceptedByUserId: json['acceptedByUserId'] as String?,
    );
  }
}
